### PR TITLE
Generate UUIDs on demand

### LINKED ISSUE #8716: Delay UUID generation in IRs
Every IR has a `UUID` assigned on creation. During compiler phases we create hundreds of copies of IRs without ever touching their ID's.
Generation of `UUID` uses _secure random_ which is expensive.

Proposal is to delay the generation of `UUID` until necessary.

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

Trying to avoid expensive `UUID.randomUUID()` unless we reallly need it.

Closes #8716.

### Important Notes

Some improvement:
Screenshot from 2024-01-11 15-16-10

FWIW Total Time for a Hello World example

Before
Screenshot from 2024-01-12 17-45-56

After
Screenshot from 2024-01-12 17-46-13

Memory usage

Before
Screenshot from 2024-01-12 17-54-54

After
Screenshot from 2024-01-12 17-54-36

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/LambdaShorthandToLambda.scala [hunk 1]
```diff
@@ -412,7 +411,7 @@ case object LambdaShorthandToLambda extends IRPass {
             )
 
         val lambdaArg = DefinitionArgument.Specified(
-          scrutineeName.copy(id = IR.randomId),
+          scrutineeName.copy(id = null),
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/LambdaShorthandToLambda.scala:414]: `null` is _"cheaper"_ value than any other value. Occupies nothing in the heap.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/CallArgument.scala [hunk 2]
```diff
@@ -49,8 +48,8 @@ object CallArgument {
     passData: MetadataStorage      = new MetadataStorage(),
     diagnostics: DiagnosticStorage = DiagnosticStorage()
   ) extends CallArgument
-      with IRKind.Primitive {
-    var id: UUID @Identifier = randomId
+      with IRKind.Primitive
+      with LazyId {
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/CallArgument.scala:52]: `LazyId` trait! That sounds promising.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package org.enso.compiler.core.ir
+
+import org.enso.compiler.core.IR.randomId
+import org.enso.compiler.core.{IR, Identifier}
+
+import java.util.UUID
+
+trait LazyId { self: IR =>
+  private[this] var _id: UUID @Identifier = null
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:9]: `private` is better than `public`!

Why do we define `self: IR` when `self` is never used? Only to ensure that `LazyId` is never applied to some other class than subclass of `IR`?

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+  override def id: UUID @Identifier = {
+    _id
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:12]: I thought that the 
```scala
    if (_id == null) {
      _id = randomId()
    }
    _id
```
would be here, in this `id` method. That this `id` should never return `null`.

Wouldn't it be getter to:
```scala
override def id: UUID @Identifier = getId()
```
?

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  override def getId(): UUID @Identifier = {
+    if (_id == null) {
+      _id = randomId()
+    }
+    _id
+  }
+
+  def id_=(id: UUID @Identifier) = _id = id
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala]: Coming from Java world, I'd call setter `setId`. But if scala needs `id_` to allow us to use `id = ...`, then OK.

In any case, I'd `assert _id == null` - e.g. let's disallow changing the ID, if it has already been provided and see whether our tests are OK.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala [hunk 6]
```diff
@@ -75,15 +75,12 @@ sealed case class Resolution(
         if (keepMetadata) passData.duplicate else new MetadataStorage(),
       diagnostics =
         if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
-      id = randomId
+      id = null
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala]: Isn't `if (keepIdentifiers)` check missing here?

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala [hunk 7]
```diff
@@ -356,31 +356,31 @@ class DataflowAnalysisTest extends CompilerTest {
...
-    val plusExprArgAExprId = mkStaticDep(plusExprArgAExpr.getId)
-    val plusExprArgBId     = mkStaticDep(plusExprArgB.getId)
-    val plusExprArgBExprId = mkStaticDep(plusExprArgBExpr.getId)
-    val frobExprId         = mkStaticDep(frobExpr.getId)
-    val frobFnId           = mkStaticDep(frobFn.getId)
-    val frobArgAId         = mkStaticDep(frobArgA.getId)
-    val frobArgAExprId     = mkStaticDep(frobArgAExpr.getId)
-    val frobArgCId         = mkStaticDep(frobArgC.getId)
-    val frobArgCExprId     = mkStaticDep(frobArgCExpr.getId)
+    val methodId           = mkStaticDep(method.getId())
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala:359]: Adding these `()` parenthesis seems like a small price to pay for avoiding `randomId` allocations.

**Review by JaroslavTulach — APPROVED** [review]
I am glad the `with LazyId` solution seems to be working. Everything is nicely green. And the startup benchmarks seem to be a bit improved too.

Can you manually check the amount of `UUID` in memory before and now at the end of execution? The memory consumption should dramatically decrease, right?

Measuring how much `UUID` has been allocated on the heap (and GCed) would be interesting as well, but I am not sure how to perform such measurement reliably.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala (hunk 3 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:9]: > Only to ensure that LazyId is never applied to some other class than subclass of IR?

Correct

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala (hunk 5 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala]: > Coming from Java world, I'd call setter setId. But if scala needs id_ to allow us to use id = ..., then OK.

Yes, that's Scala's way of saying it should be a setter. Of course I can call it `setId` but then all assignments have to be changed accordingly i.e. `id = ... ` will no longer work.

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala (hunk 7 above)
Comment (hubertp) [reply engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala:359]: lack of `()` is now discouraged in Scala. Besides the convention was that you would always use brackets if there is some side-effect.

Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala:359]: There should be no (observable) side-effect. The field is just `lazy var` - except race conditions - the code is side-effect free.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala (hunk 4 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:12]: The problem is that in IR definitions we use `.id` all over the place and we do **not** want to initialize it at that point. Only at the point when it is actually used, `.getId()`, outside of IR.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala (hunk 6 above)
Comment (hubertp) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala]: Probably, seems like this was missing before.

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala (hunk 7 above)
Comment (hubertp) [reply engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala:359]: I meant the side-effect of `var` being initialized now

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/errors/Resolution.scala]: I have found another such place last week. Yeah, sometimes there are errors in the copy/paste IR programming.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:12]: Can you make `id` _private_ then? Package private or subtree private? Or there could be:
```scala
def isIdInitialized(): Boolean
```
and we could write
```scala
id = if (keepIdentifiers && isIdInitialized()) getId() else null
```
everywhere.

File: engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/compiler/test/pass/analyse/DataflowAnalysisTest.scala:359]: That's not a side-effect as it is not observable from outside of the API. Well, if there is `id` or [isIdInitialized()](https://github.com/enso-org/enso/pull/8728#discussion_r1449109158) then it is a bit observable. Hiding these "side-effect witnessing methods" from external users of the API would be good. We don't want people outside of the `runtime-parser` project to see any side-effects.

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala (hunk 4 above)
Comment (4e6) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyId.scala:12]: If it is  only used within IRs, we can mark it _protected_

**Review by 4e6 — APPROVED** [review]

**Comment by hubertp** [comment]
Added screenshots that illustrate improvement on a simple Hello World program.

**Review by Akirathan — APPROVED** [review]
Very nice and elegant improvement. Congratulations.

**Pull request merged**